File: C++/swap-nodes-in-pairs.cpp
/**
 *  * Definition for singly-linked list.
 *   * struct ListNode {
 *    *     int val;
 *     *     ListNode *next;
 *      *     ListNode(int x) : val(x), next(NULL) {}
 *       * };
 *        */
class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        ListNode *prev, *fake, *p1, *p2;
        prev = fake = new ListNode(0);
        fake->next = head;
        while(prev->next && prev->next->next) {
            p1 = prev->next;
            p2 = p1->next;
            prev->next = p2;
            p1->next = p2->next;
            p2->next = p1;
            prev = p1;
        }
        head = fake->next;
        delete fake;
        return head;
    }
};

File: Python/swap-nodes-in-pairs.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param {ListNode} head
    # @return {ListNode}
    def swapPairs(self, head):
        prev = fake = ListNode(0)
        fake.next = head
        while prev.next and prev.next.next:
            p1 = prev.next
            p2 = p1.next
            prev.next = p2
            p1.next = p2.next
            p2.next = p1
            prev = p1
        return fake.next
